src/Services/Virtualization/PortalAuthenticationModule.cs

```diff
@@ -253,13 +187,190 @@ private void OnAuthenticateRequest(object sender, EventArgs e)
                     }
                     string message = null;
                     if (site == null)
-                        message = string.Format(HttpContext.GetGlobalResourceObject("Portal", "DefaultAuthenticationNotSupported") as string, authenticationType);
+                        message =
+                            string.Format(
+                                HttpContext.GetGlobalResourceObject("Portal", "DefaultAuthenticationNotSupported") as
+                                    string, authenticationType);
                     else
                         message = string.Format("AuthenticationNotSupportedOnSite", site.Name, authenticationType);
                     throw new NotSupportedException(message);
             }
         }
 
+        private string GetAuthenticationTypeHeader(HttpRequestBase request)
+        {
+            return request.Headers[AuthenticationTypeHeaderName] ?? request.Headers[AuthenticationTypeHeaderName.ToLower()];
+        }
+
+        private string GetRefreshHeader(HttpRequestBase request)
+        {
+            return request.Headers[RefreshHeaderName] ?? request.Headers[RefreshHeaderName.ToLower()];
+        }
+        private string GetAccessHeader(HttpRequestBase request)
+        {
+            return request.Headers[AccessHeaderName] ?? request.Headers[AccessHeaderName.ToLower()];
+        }
+
+        private bool IsTokenAuthenticationRequested(HttpRequestBase request)
+        {
+            return request.IsSecureConnection && (GetAuthenticationTypeHeader(request) == "Token"
+                || new[] {TokenLoginPath, TokenRefreshPath}.Contains(request.Url.PathAndQuery,  StringComparer.InvariantCultureIgnoreCase)

```

Comment:
url.PathAndQuery may contain query parameters (e.g. ?a=b) later for any reason, which ruins the comparison with the constants. We should compare to the path only.